File: 151051/Krupovich/330 & 340 & 350/dc_kr/src/main/java/ru/bsuir/services/LabelService.java
package ru.bsuir.services;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import lombok.AllArgsConstructor;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import ru.bsuir.dto.request.LabelRequestTo;
import ru.bsuir.dto.response.LabelResponseTo;
import ru.bsuir.entity.Label;
import ru.bsuir.irepositories.ILabelRepository;
import ru.bsuir.mapper.LabelMapper;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
@EnableCaching
public class LabelService {

    private final ILabelRepository labelRepository;
    private final LabelMapper labelMapper;

    @CacheEvict(value = "labels", allEntries = true)
    public LabelResponseTo createLabel(LabelRequestTo labelRequest) {
        Label label = labelMapper.toEntity(labelRequest);
        label = labelRepository.save(label);
        return labelMapper.toDTO(label);
    }

    @Cacheable(value = "labels", key = "#id")
    public LabelResponseTo getLabelById(Long id) {
        Optional<Label> labelOpt = labelRepository.findById(id);
        return labelOpt.map(labelMapper::toDTO)
                .orElseThrow(() -> new RuntimeException("Label not found"));
    }

    @Cacheable(value = "labelsList")
    public List<LabelResponseTo> getAllLabels(){
        return labelRepository.findAll().stream()
                .map(labelMapper::toDTO)
                .toList();
    }

    @CacheEvict(value = "labels", key = "#id")
    public LabelResponseTo updateLabel(Long id, LabelRequestTo labelRequest) {

        Label label = labelRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Label not found"));

        label.setName(labelRequest.getName());
        labelRepository.save(label);

        return labelMapper.toDTO(label);
    }

    @CacheEvict(value = "labels", key = "#id")
    public void deleteLabel(Long id) {
        if(!labelRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Label not found");
        }
        labelRepository.deleteById(id);
    }
}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/service/LabelService.kt
package bsuir.dc.publisher.service

import bsuir.dc.publisher.dto.from.LabelRequestTo
import bsuir.dc.publisher.dto.to.LabelResponseTo
import bsuir.dc.publisher.mapper.toEntity
import bsuir.dc.publisher.mapper.toResponse
import bsuir.dc.publisher.repository.IssueRepository
import bsuir.dc.publisher.repository.LabelRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.stereotype.Service

@Service
class LabelService(
    private val labelRepository: LabelRepository,
    private val issueRepository: IssueRepository,
) {
    @Caching(
        evict = [
            CacheEvict(value = ["labels"], key = "'all_labels'")
        ]
    )
    fun createLabel(labelRequestTo: LabelRequestTo): LabelResponseTo {
        val label = labelRequestTo.toEntity()
        val savedLabel = labelRepository.save(label)
        return savedLabel.toResponse()
    }

    @Cacheable(value = ["labels"], key = "#id")
    fun getLabelById(id: Long): LabelResponseTo {
        val label = labelRepository.findById(id).orElseThrow { NoSuchElementException() }
        return label.toResponse()
    }

    @Cacheable(value = ["labels"], key = "'all_labels'")
    fun getAllLabels(): List<LabelResponseTo> =
        labelRepository.findAll().map { it.toResponse() }

    @Caching(
        put = [CachePut(value = ["labels"], key = "#id")],
        evict = [CacheEvict(value = ["labels"], key = "'all_labels'")]
    )
    fun updateLabel(id: Long, labelRequestTo: LabelRequestTo): LabelResponseTo {
        val updatedLabel = labelRequestTo.toEntity().apply { this.id = id }
        return labelRepository.save(updatedLabel).toResponse()
    }

    @Caching(
        evict = [
            CacheEvict(value = ["labels"], key = "#id"),
            CacheEvict(value = ["labels"], key = "'all_labels'")
        ]
    )
    fun deleteLabel(id: Long) {
        labelRepository.findById(id).orElseThrow { NoSuchElementException() }
        labelRepository.deleteById(id)
    }

    @Cacheable(value = ["labelsByIssue"], key = "#issueId")
    fun getLabelsByIssueId(issueId: Long): List<LabelResponseTo> {
        val issue = issueRepository.findById(issueId).orElseThrow { NoSuchElementException() }
        return issue.labels.map { it.toResponse() }
    }
}